File: bad-tests-v2/declare_directive_device_resident_clause.f90
fortran
#ifndef T1
!T1:declare directive device_resident clause,V:2.7-2.3
      LOGICAL FUNCTION test1()
        USE OPENACC
        IMPLICIT NONE
        INCLUDE "acc_testsuite.Fh"
        
        INTEGER :: errors = 0
        REAL, DIMENSION(10) :: a, b, c
        REAL, DIMENSION(10) :: d, e, f
        COMMON /my_common/ a, b, c
        COMMON /my_common/ d, e, f
        
        !$acc declare device_resident(a, b, c)
        !$acc declare device_resident(d, e, f)
        
        !$acc parallel
        !$acc loop
        DO i = 1, 10
          a(i) = b(i) + c(i)
          d(i) = e(i) + f(i)
        END DO
        !$acc end parallel
        
        !$acc end declare
        
        IF (errors .eq. 0) THEN
          test1 = .FALSE.
        ELSE
          test1 = .TRUE.
        END IF
      END
#endif
      
      PROGRAM main
        IMPLICIT NONE
        INTEGER :: failcode, testrun
        LOGICAL :: failed
        INCLUDE "acc_testsuite.Fh"
#ifndef T1
        LOGICAL :: test1
#endif
        failed = .FALSE.
        failcode = 0
#ifndef T1
        DO testrun = 1, NUM_TEST_CALLS
          failed = failed .or. test1()
        END DO
        IF (failed) THEN
          failcode = failcode + 2 ** 0
          failed = .FALSE.
        END IF
#endif
        CALL EXIT (failcode)
      END PROGRAM

File: bad-tests-v2/declare_directive_device_resident_clause.cpp
#include "acc_testsuite.h"

#ifndef T1
//T1:declare directive device_resident clause,V:2.7-3.3
int test1(){
    int err = 0;
    srand(SEED);

    // Declare a variable in device memory
    int* device_var;
    device_var = (int*)acc_malloc(sizeof(int));

    // Set the value of the variable in device memory
    *device_var = 123;

    // Declare a variable in host memory
    int* host_var;
    host_var = (int*)malloc(sizeof(int));

    // Set the value of the variable in host memory
    *host_var = 456;

    // Use the device_resident clause to specify that the variable in device memory should be used
    acc_device_resident(device_var);

    // Use the device_resident clause to specify that the variable in host memory should be used
    acc_device_resident(host_var);

    // Check that the value of the variable in device memory is correct
    if (*device_var != 123){
        err = 1;
    }

    // Check that the value of the variable in host memory is correct
    if (*host_var != 456){
        err = 1;
    }

    // Free the memory allocated for the variables
    acc_free(device_var);
    free(host_var);

    return err;
}
#endif

int main(){
    int failcode = 0;
    int failed;
#ifndef T1
    failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x){
        failed = failed + test1();
    }
    if (failed != 0){
        failcode = failcode + (1 << 0);
    }
#endif
    return failcode;
}